docs/API/metadata.md

```diff
@@ -0,0 +1,121 @@
+## Dataset metadata
+
+A dataset is referenced by a formal resource ID, and also has a resource name
+for convenience. The Pbench server also maintains metadata about each dataset,
+which can help with searching and analysis. Some metadata is modifyable by an
+authenticated user, while other metadata is maintained internally by the server
+and can't be changed. Authenticated users can also add any additional metadata
+that might be of use.
+
+>__NOTE__: right now the ability to search and filter using metadata is
+limited, but our intent is to be able to use any defined metadata value both
+for searches and to filter the results of [datasets/list](V1/list.md).
+
+When a dataset is first processed, the Pbench Server will populate basic
+metadata, including the creation timestamp, the owner of the dataset (the
+username associated with the token given to the Pbench Agent
+`pbench-results-move` command), and the full contents of the dataset's
+`metadata.log` file inside the dataset tarball. These are all accessible
+under the `dataset` metadata key namespace.
+
+The Pbench server will also calculate a default deletion date for the dataset
+based on the owner's retention policy and the server administrator's retention
+policy along with some other internal management context. The expected deletion
+date is accessible under the `server` metadata key namespace.
+
+Clients can also set arbitrary metadata through the "dashboard" and "user"
+metadata namespaces:
+* The "dashboard" namespace can only be modified by the owner of the dataset,
+and is visible to anyone with read access to the dataset.
+* The "user" namespace is private to each authenticated user, and even if you
+don't own a dataset you can set your own private "user" metadata to help you
+categorize that dataset and to find it again.
```

Comment:
When somebody deletes a dataset, I would assume we would remove that user metadata?  Or would we keep it around noting that the associated dataset has been deleted?

Reply:
Yeah, it would probably be worth mentioning that, if the dataset is deleted, other users' associated metadata is deleted automatically.  (This is done by the database engine -- we don't currently have any control over this...i.e., it's a _feature_!  😆)

Reply:
It's not true that we don't have any control; rather, we've exercised that control by telling SQLAlchemy (and therefore PostgreSQL) that they're related and that deletion of the "parent" should delete all the "children".

I suppose I could say explicitly that it goes away, but that feels silly as it isn't visible to the user anyway... there's no way to get/set metadata without a dataset handle.

Reply:
I do think it's worth pointing out that, if a user creates private "user" metadata associated with a dataset owned by some _other_ user, when that dataset is deleted, **all** private "user" metadata is also deleted.  (I don't know what a user would store in "user" metadata which would be valuable after the associated dataset had been deleted, but it's probably prudent to point out this effect.)

Reply:
Again, I don't see how this is relevant at this level of description. It'd be a server bug (dangling resources) if we didn't delete them; but from the point of view of the caller, it doesn't matter as there's no possible way to access them without the dataset ID.

Reply:
Speaking most technically, which is to say, from the _interface_ perspective, I suppose it's true that there is no way for a requestor to know what happens when the dataset is deleted.  It is possible that the requestor still possesses the dataset ID, but the server will respond that the dataset does not exist, and, therefore, the fate of the metadata will be unknowable.

But, the people who are providing the UX for the end-users are going to be working based on this document.  And so, in support of their users, it would be good to make it clear that the lifetime of private "user" metadata is not entirely under the control of the user which created it, and that it will end when the dataset that it's associated with is deleted.

Reply:
I see that you've added the comment about the namespace going away down lower (at line 72); hopefully, that's sufficient.